### PR TITLE
Automatically determine CPU affinity

### PR DESCRIPTION
**Pull request by pentschev**
This PR allows automatically determining and setting CPU affinity for all GPUs during initialization.

cc @benjha as this is one of the issues we discussed on our call last week.

### CODE REVIEW

File: dask_cuda/tests/test_utils.py [hunk 1]
```diff
@@ -11,3 +13,29 @@ def test_get_n_gpus():
...
+)
+def test_bitmask_to_list(params):
+    assert bitmask_to_list(params["input"]) == params["output"]
+
+
+def test_cpu_affinity():
+    for i in range(get_n_gpus()):
+        affinity = get_cpu_affinity_list(i)
+        os.sched_setaffinity(0, affinity)
+        assert list(os.sched_getaffinity(0)) == affinity
```
Comment (mrocklin) [inline dask_cuda/tests/test_utils.py]: This seems to test the `os.sched_setaffinity`/`os.sched_getaffinity` functionality more than it tests `get_cpu_affinity_list`?

File: dask_cuda/utils.py [hunk 2]
```diff
@@ -1,6 +1,57 @@
...
+
+
+class CPUAffinity:
+    def __init__(self, cores):
+        self.cores = cores
+
+    def setup(self, worker=None):
+        print("before", os.sched_getaffinity(0))
+        os.sched_setaffinity(0, self.cores)
+        print("after ", os.sched_getaffinity(0))
```
Comment (mrocklin) [inline dask_cuda/utils.py]: We might want to remove these print statements

File: dask_cuda/utils.py [hunk 3]
```diff
@@ -1,6 +1,57 @@
...
+
+
+def bitmask_to_list(x, mask_bits=64):
```
Comment (mrocklin) [inline dask_cuda/utils.py]: Can I ask for a small docstring here, perhaps with a small example?

File: dask_cuda/utils.py [hunk 4]
```diff
@@ -1,6 +1,57 @@
...
+    return cpu_count()
+
+
+def get_cpu_affinity_list(device_index):
+    nvmlInit()
+
+    # Result is a list of 64-bit integers, thus ceil(CPU_COUNT / 64)
+    affinity = nvmlDeviceGetCpuAffinity(
+        nvmlDeviceGetHandleByIndex(device_index), math.ceil(get_cpu_count() / 64)
+    )
```
Comment (mrocklin) [inline dask_cuda/utils.py]: I'm glad to see that this function was available in nvml!

File: dask_cuda/utils.py [hunk 5]
```diff
@@ -1,6 +1,57 @@
...
+
+        bytestr = np.frombuffer(
+            bytes(np.binary_repr(mask, width=mask_bits), "utf-8"), "u1"
+        )
+        mask = np.flip(bytestr - ord("0")).astype(np.bool)
+        affinity = np.where(
+            mask, np.arange(mask_bits) + cpu_offset, np.full(mask_bits, -1)
+        )
+
+        res += affinity[(affinity >= 0)].tolist()
```
Comment (mrocklin) [inline dask_cuda/utils.py]: I like that this function is general purpose.  My guess is that terms like `affinity` here are probably leaking in from how you are using it in this case.  Maybe these should have different names?

File: dask_cuda/utils.py [hunk 6]
```diff
@@ -1,6 +1,57 @@
...
+
+    return res
+
+
+@toolz.memoize
+def get_cpu_count():
+    return cpu_count()
+
+
+def get_cpu_affinity_list(device_index):
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply dask_cuda/utils.py]: I think that adding a small docstring with a small example would also be good here.  I find that docstring examples are some of the nicest ways to communicate functionality.  I personally find them valuable when reviewing, or trying to figure out what internal code does later when debugging.

https://docs.dask.org/en/latest/develop.html#docstrings

File: dask_cuda/tests/test_utils.py (hunk 1 above)
Comment (pentschev) [reply dask_cuda/tests/test_utils.py]: Well, both. I'm actually verifying that what I'm setting is indeed workable. If I pass dummy values the result will be wrong and we can't predict the affinity of the machine running the test, so it's probably the best to check what we expected to set got set.

File: dask_cuda/utils.py (hunk 3 above)
Comment (pentschev) [reply dask_cuda/utils.py]: Renamed to `unpack_bitmask` and added docstring. I think the new name is more descriptive.

File: dask_cuda/utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pentschev) [reply dask_cuda/utils.py]: Done.

File: dask_cuda/utils.py (hunk 4 above)
Comment (pentschev) [reply dask_cuda/utils.py]: Yes, it was pretty straightforward.

File: dask_cuda/utils.py (hunk 5 above)
Comment (pentschev) [reply dask_cuda/utils.py]: Done, renamed it to `unpacked_mask`.

**Comment by pentschev** [comment]
I think all the issues have been addressed, but I think tests will still fail. I'm trying to reproduce the issue, but I can't locally. Will keep trying.

File: dask_cuda/tests/test_utils.py (hunk 1 above)
Comment (mrocklin) [reply dask_cuda/tests/test_utils.py]: Ah, good point

**Comment by mrocklin** [comment]
@rjzamora have you seen errors like this?

```python-traceback
/conda/envs/gdf/lib/python3.6/site-packages/tornado/ioloop.py:532: in run_sync

    return future_cell[0].result()

dask_cuda/tests/test_local_cuda_cluster.py:48: in test_with_subset_of_cuda_visible_devices

    scheduler_port=0, asynchronous=True, device_memory_limit=1

dask_cuda/local_cuda_cluster.py:111: in __init__

    self.scale(n_workers)

/conda/envs/gdf/lib/python3.6/site-packages/distributed/deploy/spec.py:408: in scale

    self.worker_spec.update(self.new_worker_spec())

dask_cuda/local_cuda_cluster.py:131: in new_worker_spec

    "plugins": {CPUAffinity(get_cpu_affinity(ii))},

dask_cuda/utils.py:88: in get_cpu_affinity

    nvmlDeviceGetHandleByIndex(device_index), math.ceil(get_cpu_count() / 64)

/conda/envs/gdf/lib/python3.6/site-packages/pynvml/nvml.py:922: in nvmlDeviceGetHandleByIndex

    check_return(ret)

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 



ret = 2



    def check_return(ret):

        if (ret != NVML_SUCCESS):

>           raise NVMLError(ret)

E           pynvml.nvml.NVMLError_InvalidArgument: Invalid Argument

```

**Comment by pentschev** [comment]
I understand the error now. It means the device with the index `ii` doesn't exist. This is because on CI there's only one GPU available, but we still treat it as if there were several. I'll probably need to rework that test.

**Comment by pentschev** [comment]
The issue is fixed, but there was an error on downloading packages, I'm rerunning the tests just in case.

**Comment by pentschev** [comment]
rerun tests

**Comment by pentschev** [comment]
All good for a new review now @mrocklin.

**Comment by mrocklin** [comment]
This looks great to me.  Thanks @pentschev !

**Pull request merged**